File: Implementierung/Backend/src/main/java/edu/kit/informatik/service/SeatArrangementService.java
package edu.kit.informatik.service;

import edu.kit.informatik.dto.mapper.courses.SeatArrangementMapper;
import edu.kit.informatik.dto.userdata.courses.SeatArrangementDto;
import edu.kit.informatik.exceptions.EntityNotFoundException;
import edu.kit.informatik.model.userdata.courses.Course;
import edu.kit.informatik.model.userdata.courses.SeatArrangement;
import edu.kit.informatik.repositories.CourseRepository;
import edu.kit.informatik.repositories.SeatArrangementRepository;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.security.core.Authentication;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;

/**
 * Service für {@link SeatArrangement Sitzordnungen}.
 *
 * @author ugqbo
 * @version 1.0
 */

@Component
public class SeatArrangementService extends BaseService<SeatArrangement, SeatArrangementDto, SeatArrangementDto> {

    private static final String ID_ATTRIBUTE = "userId";

    private final SeatArrangementRepository seatArrangementRepository;
    private final CourseRepository courseRepository;

    /**
     * Konstruktor zum Erstellen eines Objektes der Klasse
     *
     * @param seatArrangementRepository {@link SeatArrangementRepository}
     * @param seatArrangementMapper     {@link SeatArrangementMapper}
     * @param courseRepository          {@link CourseRepository}
     */
    public SeatArrangementService(SeatArrangementRepository seatArrangementRepository,
                                  SeatArrangementMapper seatArrangementMapper, CourseRepository courseRepository) {
        super(seatArrangementMapper);
        this.seatArrangementRepository = seatArrangementRepository;
        this.courseRepository = courseRepository;
    }

    @Override
    public SeatArrangementDto add(SeatArrangementDto seatArrangementDto, Authentication authentication) {
        super.checkAuthorization(authentication, seatArrangementDto.getUserId());
        SeatArrangement seatArrangement = this.mapper.dtoToModel(seatArrangementDto);
        SeatArrangement newSeatArrangement = this.seatArrangementRepository.save(seatArrangement);
        return this.mapper.modelToDto(newSeatArrangement);
    }

    @Transactional
    @Override
    public SeatArrangementDto update(SeatArrangementDto seatArrangementDto, Authentication authentication) {
        super.checkAuthorization(authentication, seatArrangementDto.getUserId());
        Optional<SeatArrangement> repositorySeatArrangementOptional = this.seatArrangementRepository
                                                                .findSeatArrangementById(seatArrangementDto.getId());
        
        SeatArrangement repositorySeatArrangement = repositorySeatArrangementOptional.orElseThrow(
                        () -> new EntityNotFoundException(SeatArrangement.class, seatArrangementDto.getId()));
        SeatArrangement newSeatArrangement = this.mapper.dtoToModel(seatArrangementDto);

        if (!newSeatArrangement.getName().equals(repositorySeatArrangement.getName())) {
            repositorySeatArrangement.setName(newSeatArrangement.getName());
        }
        if (!newSeatArrangement.getSeatMap().equals(repositorySeatArrangement.getSeatMap())) {
            repositorySeatArrangement.setSeatMap(newSeatArrangement.getSeatMap());
        }
        
        return mapper.modelToDto(repositorySeatArrangement);
    }

    @Override
    public SeatArrangementDto getById(String id, Authentication authentication) {
        Optional<SeatArrangement> seatArrangementOptional = this.seatArrangementRepository.findSeatArrangementById(id);
        SeatArrangement seatArrangement = seatArrangementOptional.orElseThrow(() ->
                                                        new EntityNotFoundException(SeatArrangement.class, id));
        super.checkAuthorization(authentication, seatArrangement.getUser().getId());


        return this.mapper.modelToDto(seatArrangement);
    }

    @Override
    public List<SeatArrangementDto> getAll(Authentication authentication) {
        JwtAuthenticationToken jAT = (JwtAuthenticationToken) authentication;

        return mapper.modelToDto(this.seatArrangementRepository.findSeatArrangementsByUserId(
                                                                jAT.getTokenAttributes().get(ID_ATTRIBUTE).toString()));
    }

    @Override
    public String delete(String id, Authentication authentication) {
        Optional<SeatArrangement> seatArrangementOptional = this.seatArrangementRepository.findSeatArrangementById(id);
        SeatArrangement seatArrangement = seatArrangementOptional.orElseThrow(() ->
                                                            new EntityNotFoundException(SeatArrangement.class, id));
        super.checkAuthorization(authentication, seatArrangement.getUser().getId());

        return delete(seatArrangement);
    }

    /**
     * Methode zum Löschen einer {@link SeatArrangement}
     * @param seatArrangement {@link SeatArrangement}
     * @return Id des {@link SeatArrangement}
     */
    protected String delete(SeatArrangement seatArrangement) {
        Course course = courseRepository.findCourseBySeatArrangements(seatArrangement);
        if (course != null) {
            course.getSeatArrangements().remove(seatArrangement);
        }

        this.seatArrangementRepository.deleteById(seatArrangement.getId());

        return seatArrangement.getId();
    }
}

File: Implementierung/frontend/paschi/src/service/SeatArrangementService.ts
import {BASE_URL, BaseService} from "@/service/BaseService";
import {SeatArrangement} from "@/model/userdata/courses/SeatArrangement";
import {SeatArrangementDto} from "@/dto/userdata/courses/SeatArrangementDto";
import {SeatArrangementMapper} from "@/dto/mapper/courses/SeatArrangementMapper";
import axios, {AxiosResponse} from "axios";
import {useUserStore} from "@/store/UserStore";

const SEAT_ARRANGEMENT_BASE_URL: string = BASE_URL + "/api/seatarrangement";

export class SeatArrangementService extends BaseService<SeatArrangement, SeatArrangementDto> {

  private static seatArrangementService: SeatArrangementService = new SeatArrangementService();

  private constructor() {
    super(SeatArrangementMapper.getMapper());
  }

  static getService(): SeatArrangementService {
    return this.seatArrangementService;
  }

  async add(arrangement: SeatArrangement) {
    const token = useUserStore().getUser()?.token;
    const arrangementDto = this.getMapper().modelToDto(arrangement);
    await axios
      .post(SEAT_ARRANGEMENT_BASE_URL, arrangementDto, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .then((response: AxiosResponse<SeatArrangementDto>) => {
        arrangement.setId = response.data.id;
      })
      .catch((error) => {
        console.log(error);
      });
  }

  async update(arrangement: SeatArrangement) {
    const token = useUserStore().getUser()?.token;
    const arrangementDto = this.getMapper().modelToDto(arrangement);
    await axios
      .put(SEAT_ARRANGEMENT_BASE_URL, arrangementDto, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .catch((error) => {
        console.log(error);
      });
  }

  async getById(id: string): Promise<SeatArrangement | undefined> {
    const token = useUserStore().getUser()?.token;
    let arrangement;
    await axios
      .get(SEAT_ARRANGEMENT_BASE_URL + `/${id}`, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .then(async (response: AxiosResponse<SeatArrangementDto>) => {
        arrangement = await this.getMapper().dtoToModel(response.data);
      })
      .catch((error) => {
        console.log(error);
      });

    return arrangement;
  }

  async getAll(): Promise<SeatArrangement[]> {
    const token = useUserStore().getUser()?.token;
    const arrangements: SeatArrangement[] = [];
    await axios
      .get(SEAT_ARRANGEMENT_BASE_URL, {
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .then(async (response: AxiosResponse<SeatArrangementDto[]>) => {
        for (const arrangementDto of response.data) {
          arrangements.push(await this.getMapper().dtoToModel(arrangementDto));
        }
      })
      .catch((error) => {
        console.log(error);
      });

    return arrangements;
  }

  async delete(id: string) {
    const token = useUserStore().getUser()?.token;
    await axios
      .delete(SEAT_ARRANGEMENT_BASE_URL, {
        params: {
          id,
        },
        headers: {
          Authorization: `Bearer ${token}`,
        },
      })
      .catch((error) => {
        console.log(error);
      });
  }
}
